src/background_workers.c

```diff
@@ -0,0 +1,197 @@
+  int result = REDISMODULE_ERR;
+  if (run_queues == NULL) {
+    return result;
+  }
+
+  AI_dictEntry *entry = AI_dictFind(run_queues, devicestr);
+  if (entry) {
+    *run_queue_info = AI_dictGetVal(entry);
+    result = REDISMODULE_OK;
+  } else {
+    *run_queue_info = RedisModule_Alloc(sizeof(RunQueueInfo));
+    (*run_queue_info)->run_queue = queueCreate();
+    pthread_cond_init(&(*run_queue_info)->queue_condition_var, NULL);
+    pthread_mutex_init(&(*run_queue_info)->run_queue_mutex, NULL);
+    (*run_queue_info)->threads = (pthread_t *)RedisModule_Alloc(
+        sizeof(pthread_t) * perqueueThreadPoolSize);
+    /* create threads */
+    for (int i = 0; i < perqueueThreadPoolSize; i++) {
+      if (pthread_create(&((*run_queue_info)->threads[i]), NULL,
+                         RedisAI_Run_ThreadMain, *run_queue_info) != 0) {
+        freeRunQueueInfo(*run_queue_info);
+        return REDISMODULE_ERR;
+      }
+    }
+    AI_dictAdd(run_queues, (void *)devicestr, (void *)*run_queue_info);
+    result = REDISMODULE_OK;
+  }
+
+  return result;
+}
+
+void *RedisAI_Run_ThreadMain(void *arg) {
+  RunQueueInfo *run_queue_info = (RunQueueInfo *)arg;
+  pthread_t self = pthread_self();
+#ifdef __APPLE__
+  int res = pthread_setname_np("redisai_bthread");
+#else
+  int res = pthread_setname_np(self, "redisai_bthread");
+#endif
+  pthread_mutex_lock(&run_queue_info->run_queue_mutex);
+  while (true) {
+    int rc = pthread_cond_wait(&run_queue_info->queue_condition_var,
+                               &run_queue_info->run_queue_mutex);
+
+    long long run_queue_len = queueLength(run_queue_info->run_queue);
+
+    while (run_queue_len > 0) {
+      queueItem **evicted_items = NULL;
+      RedisAI_RunInfo **batch_rinfo = NULL;
+
+      queueItem *item = queueFront(run_queue_info->run_queue);
+
+      while (item) {
+        RedisAI_RunInfo *rinfo = (RedisAI_RunInfo *)item->value;
+
+        if (evicted_items) {
+          array_free(evicted_items);
+          array_free(batch_rinfo);
+        }
+        evicted_items = array_new(queueItem *, run_queue_len);
+        batch_rinfo = array_new(RedisAI_RunInfo *, run_queue_len);
+
+        array_append(evicted_items, item);
+        array_append(batch_rinfo, rinfo);
+
+        if (rinfo->sctx) {
+          break;
+        }
+
+        // DAGRUN
+        if (rinfo->use_local_context==1){
+          break;
+        }
+
+        size_t batchsize = rinfo->mctx->model->opts.batchsize;
+
+        if (batchsize == 0) {
+          break;
+        }
+
+        size_t current_batchsize = RAI_RunInfoBatchSize(rinfo);
+
+        if (current_batchsize == 0 || current_batchsize >= batchsize) {
+          break;
+        }
+
+        queueItem *next_item = item->next;
+
+        while (next_item != NULL) {
+          RedisAI_RunInfo *next_rinfo = (RedisAI_RunInfo *)next_item->value;
+
+          if (RAI_RunInfoBatchable(rinfo, next_rinfo) == 0) {
```

Comment:
Here we need to prevent a DAG run info to be considered batchable. I'm pushing a commit to make that happen.

Reply:
BTW, we should note this limitation and extend batching to DAG as well at some point. This will require some DAG refactoring: individual MODELRUN commands should be placed in the queue, so they have a chance to be batched. We'll have to make batching logic slightly more complicated to avoid the order of DAG commands to change for an individual client due to batching.
We can leave this out for now, but I'll open an issue.